src/native/libs/System.Native/pal_networking.c

```diff
@@ -1588,6 +1588,50 @@ int32_t SystemNative_Connect(intptr_t socket, uint8_t* socketAddress, int32_t so
     return err == 0 ? Error_SUCCESS : SystemNative_ConvertErrorPlatformToPal(errno);
 }
 
+int32_t SystemNative_Connectx(intptr_t socket, uint8_t* socketAddress, int32_t socketAddressLen, uint8_t* data, int32_t dataLen, int32_t tfo, int* sent)
+{
+    if (socketAddress == NULL || socketAddressLen < 0 || sent == NULL)
+    {
+        return Error_EFAULT;
+    }
+
+    int fd = ToFileDescriptor(socket);
+#if HAVE_CONNECTX
+    struct sa_endpoints eps;
+    struct iovec iovec;
+    int flags = 0;
+    iovec.iov_base = data;
+    iovec.iov_len = dataLen;
+    memset(&eps, 0, sizeof(eps));
+    eps.sae_dstaddr = (struct sockaddr *)socketAddress;
+    eps.sae_dstaddrlen = socketAddressLen;
+
+    size_t length = 0;
+    int err;
+    while ((err = connectx(fd, &eps, SAE_ASSOCID_ANY, tfo != 0 ? CONNECT_DATA_IDEMPOTENT : 0, dataLen > 0 ? &iovec : NULL, dataLen > 0 ? 1 : 0, &length, NULL)) < 0 && errno == EINTR);
+    *sent = (int)length;
+
+    return err == 0 ? Error_SUCCESS : SystemNative_ConvertErrorPlatformToPal(errno);
+#else
+#ifdef TCP_FASTOPEN_CONNECT
+    int enabled = 1;
+    socklen_t len = sizeof(enabled);
+
+    if (tfo)
+    {
+        // This will either success and connect will finish without sending SYN until we write to so the socket.
+        // If this is not available we simply connect and write provided data afterwards.
+        setsockopt(fd, IPPROTO_TCP, TCP_FASTOPEN_CONNECT, &enabled, len);
```

Comment:
I don't see this code path sending out the data.

From the [man pages](https://man7.org/linux/man-pages/man7/tcp.7.html)><

> If a cookie is not available (i.e. first contact to the
> destination), connect(2) behaves as usual by sending a SYN
> immediately, except the SYN would include an empty Fast
> Open cookie option to solicit a cookie.
> 
> If a cookie is available, connect(2) would return 0
> immediately but the SYN transmission is deferred.  A
> subsequent write(2) or sendmsg(2) would trigger a SYN with
> data plus cookie in the Fast Open option.  In other words,
> the actual connect operation is deferred until data is
> supplied.

Reply:
there is no else. if TCP_FASTOPEN_CONNECT is not available it would simply fall-back to option 2).  e.g. be the same as if the cookie is missing.

Reply:
I meant the branch that follows

```
#else
#ifdef TCP_FASTOPEN_CONNECT
```

Isn't this supposed to work with `saea.SetBuffer(...); socket.ConnectAsync(saea);` on all platforms?
There needs to be a `sendto(...)` to make that work on Linux.

Reply:
Nevermind this, I see now that this is the responsibility of the managed `ConnectOperation` logic.

Reply:
yes, but not here. The method reports back how many bytes were sent. It may be lest than asked for. In case of Linux, it is always 0. `ConnectAsync` will check and if there are remaining bytes to sent it will use `SendToAsync` to deal with retires and async.  

Without it the old and new tests would not pass.